File: python/程序员面试金典/面试题 17.12. BiNode.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def convertBiNode(self, root: TreeNode) -> TreeNode:
        self.pre = self.ans = TreeNode(-1)
        def dfs(root):
            if not root: return
            dfs(root.left)

            root.left = None
            self.pre.right = root
            self.pre = root

            dfs(root.right)
        dfs(root)
        return self.ans.right

File: go/general/面试题 17.12. BiNode.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func convertBiNode(root *TreeNode) *TreeNode {
	pre := &TreeNode{}
	head := pre
	dfs(root, pre)
	return head.Right
}

func dfs(root *TreeNode, pre *TreeNode) *TreeNode {
	if root == nil {
		return pre
	}
	pre = dfs(root.Left, pre)
	root.Left = nil
	pre.Right = root
	pre = root
	pre = dfs(root.Right, pre)
	return pre
}
